xi-win-ui/src/widget/flex.rs

```diff
@@ -61,31 +85,81 @@ impl Axis {
 impl Row {
     pub fn new() -> Flex {
         Flex {
+            params: BTreeMap::new(),
             direction: Axis::Horizontal,
 
+            phase: Phase::NonFlex,
             ix: 0,
-            major_per_flex: 0.0,
             minor: 0.0,
+            total_non_flex: 0.0,
+            flex_sum: 0.0,
         }
     }
 }
 
 impl Column {
     pub fn new() -> Flex {
         Flex {
+            params: BTreeMap::new(),
             direction: Axis::Vertical,
 
+            phase: Phase::NonFlex,
             ix: 0,
-            major_per_flex: 0.0,
             minor: 0.0,
+            total_non_flex: 0.0,
+            flex_sum: 0.0,
         }
     }
 }
 
 impl Flex {
+    /// Add to UI with children.
     pub fn ui(self, children: &[Id], ctx: &mut UiInner) -> Id {
         ctx.add(self, children)
     }
+
+    /// Set the flex for a child widget.
+    ///
+    /// This function is used to set flex for a child widget, and is done while
+    /// building, before adding to the UI. Likely we will need to think of other
+    /// mechanisms to change parameters dynamically after building.
+    pub fn set_flex(&mut self, child: Id, flex: f32) {
+        let params = self.get_params_mut(child);
+        params.flex = flex;
+    }
+
+    fn get_params_mut(&mut self, child: Id) -> &mut Params {
+        self.params.entry(child).or_default()
+    }
+
+    fn get_params(&self, child: Id) -> Params {
+        self.params.get(&child).cloned().unwrap_or(Default::default())
+    }
+
+    /// Return the index (within `children`) of the next child that belongs in
+    /// the specified phase.
+    fn get_next_child(&self, children: &[Id], start: usize, phase: Phase) -> Option<usize> {
+        for ix in start..children.len() {
+            if (self.get_params(children[ix]).flex == 0.0) == (phase == Phase::NonFlex) {
```

Comment:
I had to stop here and think this statement through. While elegant, I wonder if it's more readable for casual readers like me, if we do instead
```rust
if (self.get_params(children[ix]).flex == 0.0 && phase == Phase::NonFlex) ||
   (self.get_params(children[ix]).flex > 0.0 && phase == Phase::Flex) {
    return Some(ix);
}
```
or introduce a `fn verify_phase(flex: f32, phase: Phase) -> bool`, assuming it is an invariant that flex == 0.0 mens it's NonFlex.